### ISSUE TITLE
Incorrect handling of CPP and CPPFLAGS in XSDKCompilerDetect.cmake

### ISSUE THREAD

**Issue by besserox** [precice/precice#756]
**Describe your setup**

This bug was initially discovered when building on our HPC platform (CentOS via EasyBuild), but it can be reproduced on _Ubuntu 19.10_ and probably any other systems.
preCICE Version: 2.0.2

**Describe the problem**

CMake fails when trying to set `CPP` and `CPPFLAGS` under certain conditions.
There are actually 2 bugs:
- `CPP` is a C PreProcessor, not a C++ compiler, so it cannot be used a such
- When environment variable `CPPFLAGS` and CMake option `CMAKE_CXX_FLAGS` are both defined, `list(APPEND)` inserts a `;` in the list of flags (because that's how list works in CMake) which is then interpreted a 2 different commands when CMake tests the C++ compiler.

**Step To Reproduce**

For the `CPP` bug:
```
CPP=cpp cmake ..
```
will just hang because `cpp` waits for the standard input.

For the `CPPFLAGS` bug:
```
CPPFLAGS=-DTEST cmake .. -DCMAKE_CXX_FLAGS='-Wall'
```
will fail with
```
    Building CXX object CMakeFiles/cmTC_b4c75.dir/testCXXCompiler.cxx.o
    /usr/bin/c++    -Wall;-DTEST    -o CMakeFiles/cmTC_b4c75.dir/testCXXCompiler.cxx.o -c /home/xavier/work/precice/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    c++: fatal error: no input files
    compilation terminated.
    /bin/sh: 1: -DTEST: not found
```

**Discussion**

I believe you want to follow the xSDK Community Installation Policies and take into account these flags, but CMake does not support them and I'm not sure there is an easy workaround.
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/3633

For `CPP`, I believe CMake does not support it because it would mean adding a additional step for every compilation of a file, ie first run CPP, and pass the output to CXX to compile it.

For `CPPFLAGS`, it is not that easy to get it correct:
- `list(APPEND)` does not work for the reason above
- `set(CMAKE_CXX_FLAGS "..."` is ignored if the value is already set by the user call CMake
- `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} $ENV{CPPFLAGS}" CACHE STRING "" FORCE)` will work but the `CPPFLAGS` will keep being added everytime CMake is run.
- Maybe a better logic could be added to get it right.

My feeling is that this issue is not for preCICE to solve, but either:
- by CMake (but for `CPP`, I am not sure it will happen)
- by having xSDK relaxing policies, in particular because CMake is one of their preferred build tool.

How do the other CMake-based xSDK-compatible software handle this?

As an alternative, I could suggest displaying a warning in case one of these variables is defined to inform the user it will not be used.

**Comment by fsimonis** [comment]
@besserox Thank you for opening the issue and doing some research on the topic.
There are many interesting aspects to this issue, so let me tackle them one at the time.

## `CPP` and `CPPFLAGS`

There are 2 inconsistent uses for these flags:
1. they specify the C/C++ preprocessor and flags. [GNU Make](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html)
2. they are used as synonyms for `CXX` and `CXXFLAGS` [xSDK](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/3.md)

## The xSDK Installation Policies

The [installation policy 3](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/3.md) requires to detect `CPP` and `CPPFLAGS` as compiler and compiler flags, thus requiring to respect the convention Nr 2 mentioned above.

There is an open discussion going on whether to remove a majority of the installation policies. The majority became obsolete due to the adaption of spack as their package/dependency manager. https://github.com/xsdk-project/xsdk-community-policies/pull/20

## Native CMake Support

After I implemented the installation policy 3, I opened the [merge request](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/3633) you mentioned, which was closed.

They came to the conclusion that following convention 1 is the way to go. More on the rational [here](https://gitlab.kitware.com/cmake/cmake/issues/12928).
The challenges for a [CMake-native solution](https://gitlab.kitware.com/cmake/cmake/issues/12928#note_757148) were listed a few minutes ago.

---

Back to your issue:

> My feeling is that this issue is not for preCICE to solve

I totally agree with you!
That said, I am reluctant to remove the support as long as the detection is required by xSDK.

> How do the other CMake-based xSDK-compatible software handle this?

My first pick (deal.ii) does not support the required flags. So, this seems to be a bit of a grey area.

I think the only reasonable way forward is to remove the functionality and hope for proper support from CMake. Removing this in develop should not be a problem as the requirement will most likely be dropped anyhow.

**Comment by besserox** [comment]
### About the xSDK Installation Policy

I believe you're mis-interpreting the [installation policy 3](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/3.md). It does not say the `CPP` (resp. `CPPFLAGS`) is a synonym for `CXX` (resp. `CXXFLAGS`). 

It is true that this policy calls `CPP` and `CPPFLAGS`  *compiler* and *compiler flags*, but I believe it is just unfortunate shortcut / misuse of language. And I am pretty sure it is not what they meant. Also, it does not make sense to change the semantic defined in the GNU Make documentation. If you have some doubt, I would recommend to ask the xSDK community some clarification about that (and to update their documentation).

### About the xSDK Installation Policy support in preCICE

> That said, I am reluctant to remove the support as long as the detection is required by xSDK.

I recommend to remove it because it breaks the build when used:
- About `CPP`, please refer to my initial example.
- About `CPPFLAGS`, there is my initial example, but I also came up with this new one:
```
export CPPFLAGS="-DTEST"
cmake ..                       # First CMake run works
touch ../CMakeLists.xt
make                           # Second CMake run fails, not sure why
```
This example works correctly if setting `CXXFLAGS` instead of `CPPFLAGS`.

### About the Native CMake Support

> They came to the conclusion that following convention 1 is the way to go. More on the rational here. The challenges for a CMake-native solution were listed a few minutes ago.

The CMake support for `CPPFLAGS` would solve half of your problem.

However, I believe CMake will not support `CPP` because it forces to create an additional step in the compilation of each file (for almost no benefit). For example, with GCC, it is possible to run the C PreProcessor manually beforehand and pass it to GCC, but AFAIK, it is not possible to tell GCC to use a different C PreProcessor. 

### Bonus issue

When not defined, you're forcing the value of `CMAKE_BUILD_TYPE` to `Debug`. I believe it breaks the point 3.iii of the [installation policy 3](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/3.md), because if some `CXXFLAGS` are already defined, the CMake Debug flags will still be added to the `CXXFLAGS` specified by the user.



In conclusion, that's not an easy task :-)

**Comment by fsimonis** [comment]
### About the xSDK Installation Policy

> It is true that this policy calls CPP and CPPFLAGS compiler and compiler flags, but I believe it is just unfortunate shortcut / misuse of language.

I see your point. This policy definitely suffers from an overly short description.

### About the Native CMake Support

A possible future support aside, as a library, preCICE has to stay compatible with other distros. The price to pay is that we have to stick to an older minimum CMake version. Our current baseline is Ubuntu 18.04 LTS with [cmake 3.10.2](https://packages.ubuntu.com/bionic/cmake). Thus, future we cannot rely on this support.

### Bonus issue

This is also a tricky one.
[Inst. Pol. 1](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/1.md) allows packages to use the xSDK defaults as general defaults, meaning that [Inst. Pol. 4](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/4.md) defines `CMAKE_BUILD_TYPE=Debug` as default, which influences the final C++ flags in addition to the ones provided by [Inst. Pol. 3](https://github.com/xsdk-project/xsdk-community-policies/blob/master/installation_policies/3.md).
Thus I expect that this not a violation.

### To conclude this issue:

Context:

* the support of `CPP` and `CPPFLAGS` should be handled directly by CMake ( [GitLab issue](https://gitlab.kitware.com/cmake/cmake/issues/12928#note_757148) )
* the wording of the installation policy is unclear in some aspects
* the xSDK installation policy will most likely be removed as the adoption of spack makes the majority of those policies obsolete (  https://github.com/xsdk-project/xsdk-community-policies/pull/20 )
* the current implementation does not work reliably

Plan of action:

* We will remove the detection of `CPP` and `CPPFLAGS` in #757
